### ISSUE TITLE
`unused-io-amount` should not be emitted for single-byte write

### ISSUE THREAD

**Issue by mmastrac** [rust-lang/rust-clippy#14583]
### Summary

Open-source code available here: https://github.com/progscrape/progscrape

If writing a single-byte slice, `unused-io-amount` should not be emitted, as write will only return 1 or Err in this case:

```
        const NEWLINE: [u8; 1] = [b'\n'];
        let mut earliest = StoryDate::MAX;
        let mut latest = StoryDate::MIN;
        let mut count = 0;
        scrapes.fetch_all(
            shard,
            |scrape| {
                count += 1;
                earliest = earliest.min(scrape.date);
                latest = latest.max(scrape.date);
                w.write_all(serde_json::to_string(&scrape)?.as_bytes())?;
                w.write(&NEWLINE)?;
                Ok(())
            },
            |error| {
                tracing::error!("Error fetching scrape: {:?}", error);
            },
        )?;

```

```
warning: written amount is not handled
  --> application/src/persist/backerupper.rs:86:17
   |
86 |                 w.write(&NEWLINE)?;
   |                 ^^^^^^^^^^^^^^^^^^
   |
   = help: use `Write::write_all` instead, or handle partial writes
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_io_amount
   = note: `-D clippy::unused-io-amount` implied by `-D clippy::all`
   = help: to override `-D clippy::all` add `#[allow(clippy::unused_io_amount)]`
```

### Lint Name

unused-io-amount

### Reproducer

I tried this code:

```rust
        // Write each scrape to the file, with a newline separating them
        let mut w = BufWriter::new(file);
        const NEWLINE: [u8; 1] = [b'\n'];
        let mut earliest = StoryDate::MAX;
        let mut latest = StoryDate::MIN;
        let mut count = 0;
        scrapes.fetch_all(
            shard,
            |scrape| {
                count += 1;
                earliest = earliest.min(scrape.date);
                latest = latest.max(scrape.date);
                w.write_all(serde_json::to_string(&scrape)?.as_bytes())?;
                w.write(&NEWLINE)?;
                Ok(())
            },
            |error| {
                tracing::error!("Error fetching scrape: {:?}", error);
            },
        )?;

```

I saw this happen:

```
warning: written amount is not handled
  --> application/src/persist/backerupper.rs:86:17
   |
86 |                 w.write(&NEWLINE)?;
   |                 ^^^^^^^^^^^^^^^^^^
   |
   = help: use `Write::write_all` instead, or handle partial writes
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_io_amount
   = note: `-D clippy::unused-io-amount` implied by `-D clippy::all`
   = help: to override `-D clippy::all` add `#[allow(clippy::unused_io_amount)]`
```

I expected to see this happen:

No warning.

### Version

```text

```

**Comment by mmastrac** [comment]
> The [documentation of `std::io::Write::write`](https://doc.rust-lang.org/std/io/trait.Write.html#tymethod.write) doesn't seem to agree with this claim: _A return value of `Ok(0)` typically means that the underlying object is no longer able to accept bytes and will likely not be able to in the future as well, or that the buffer provided is empty._

Fair. I can't come up with a genuine, realistic implementation of write that would validly return zero, but there's a very small possibility that someone might be writing to a broken special file on Linux with a kernel implementation that incorrectly returns 0 rather than a EBUSY or EAGAIN error.
